fms_extras/utils/cache/__init__.py

```diff
@@ -0,0 +1,245 @@
+import abc
+import dataclasses
+from typing import List, Optional, Tuple
+
+import torch
+
+
+def compute_position_ids(
+    num_tokens_per_sequence: List[int], context_lengths: Optional[List[int]] = None
+) -> List[List[int]]:
+    """Compute position ids based on the current context lengths and the new tokens to add
+
+    Parameters
+    ----------
+    num_tokens_per_sequence: List[int]
+        number of tokens to be added to each sequence
+    context_lengths: List[int], optional
+        optional list of current context lengths per sequence. If none, will assume no context length and starting
+        position will be 0 (default is None)
+
+    Returns
+    -------
+    List[List[int]]
+        the position ids for each sequence
+    """
+
+    max_tokens = max(num_tokens_per_sequence)
+    position_ids = []
+    for seq_i, num_tokens in enumerate(num_tokens_per_sequence):
+        if context_lengths is None:
+            start = 0
+        else:
+            start = context_lengths[seq_i] - 1
+        position_ids_i = [0 for _ in range(max_tokens - num_tokens)] + [
+            i for i in range(start, start + num_tokens)
+        ]
+        position_ids.append(position_ids_i)
+    return position_ids
+
+
+class AttentionComputationMixin(metaclass=abc.ABCMeta):
```

Comment:
seems this type isn't really used? one class mixes it in, but I'm not seeing polymorphism, and the actual switching between attention mechanisms is done with conditional statements.

Reply:
This was left over from when we were going to implement a version of flash computation, or if another form comes along and we want to use the same MHA. Because I kept this PR to only `paged`, we probably don't need this for this PR. Having said that, if we ever use these structures for other forms of attention in the future, it might make sense to keep it

Reply:
I think we should probably remove until/unless we have a use-case for it. it would be easy to re-add when/if we decide to build something that way

Reply:
Ok will do